### PR TITLE
[Tree assume] Add `BodyContains`

### PR DESCRIPTION
**Pull request by rtjoa**
Add `IsValid` to tree-assume. See #226.

---

PR stack:
1. #336
2. #337 
3. #340 
4. #338 
3. #339

### CODE REVIEW

**Review by oflatt — CHANGES_REQUESTED** [review]
Request for a test or two more

File: tree_assume/src/optimizations/body_contains.rs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        num1 = int(1),
+        num3 = int(3),
+        num4 = int(4),
+        num5 = int(5),
+        tup45 = parallel!(int(4), int(5)),
+    );
+    crate::egglog_test(
+        &build,
+        &check,
+        vec![],
```
[the remark this answers is not in the document]
Comment (rtjoa) [reply tree_assume/src/optimizations/body_contains.rs:94]: Added

File: tree_assume/src/optimizations/body_contains.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        .chain(Constructor::iter().filter_map(captured_expr_rule_for_ctor))
+        .chain(Constructor::iter().filter_map(subexpr_rule_for_ctor))
+        .collect::<Vec<_>>()
+}
+
+#[cfg(test)]
+use crate::ast::*;
+#[cfg(test)]
+use crate::Value;
+#[test]
```
[the remark this answers is not in the document]
Comment (rtjoa) [reply tree_assume/src/optimizations/body_contains.rs:63]: Added an inner let

File: tree_assume/src/optimizations/body_contains.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (oflatt) [reply tree_assume/src/optimizations/body_contains.rs:94]: Did you forget to push this or is my diff wrong?
We might want egglog_test to panic on an empty vector

**Review by oflatt — CHANGES_REQUESTED** [review]

File: tree_assume/src/optimizations/body_contains.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (rtjoa) [reply tree_assume/src/optimizations/body_contains.rs]: Oops, actually added

**Review by oflatt — APPROVED** [review]

**Pull request merged**